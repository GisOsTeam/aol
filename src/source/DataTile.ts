import { GeoTIFF } from './GeoTIFF';

export type DataTile = GeoTIFF;
