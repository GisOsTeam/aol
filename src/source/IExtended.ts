import OlMap from 'ol/Map';
import Source, { Options } from 'ol/source/Source';
import Feature from 'ol/Feature';
import Geometry from 'ol/geom/Geometry';
import Projection from 'ol/proj/Projection';
import { SourceType } from './types/sourceType';
import { LayerType } from './types/layerType';
import { AndPre, IPredicate, Or } from '../filter/predicate';
import { FilterBuilder } from '../filter/FilterBuilder';

export interface ISnapshotOptions extends Options {
  snapshotable?: boolean;
  listable?: boolean;
  removable?: boolean;
}

export interface ISnapshotSource extends Source {
  getSourceType(): SourceType;

  getSourceOptions(): ISnapshotOptions;

  setSourceOptions(options: ISnapshotOptions): void;

  getLayerType(): LayerType;

  isSnapshotable(): boolean; // Can save snapshot
  isListable(): boolean; // Visible by others tools
  isRemovable(): boolean; // Removable from map by others tools
}

export interface IInitSource extends ISnapshotSource {
  init(): Promise<void>;
}

export interface IQuerySource extends ISnapshotSource {
  query(identifyRequest: IGisRequest, onlyVisible?: boolean): Promise<IQueryResponse>;

  retrieveFeature(id: number | string, featureProjection: Projection): Promise<Feature>;
}

export interface ILayerLegend {
  /**
   * Label optinal
   */
  label?: string;
  /**
   * Image element
   */
  image: HTMLImageElement;
  /**
   * Image data in Base64 ou url
   */
  srcImage: string;
  /**
   * Height
   */
  height: number;
  /**
   * Width
   */
  width: number;
}

export type ILegendRecord = Record<number | string, ILayerLegend[]>;

export interface ILegendSource {
  fetchLegend(): Promise<ILegendRecord>;
}

export interface IExtended extends IInitSource, IQuerySource, ILegendSource { }

export type IGisRequest = IQueryRequest | IIdentifyRequest;

export interface IAbstractRequest<T extends string> {
  olMap: OlMap;
  queryType: T;
  filters?: IPredicate;
  limit?: number;
  returnGeometry?: boolean;
  srId?: string;
}

export interface IQueryRequest extends IAbstractRequest<'query'> {
  geometry?: Geometry;
  geometryProjection?: Projection;
  geometryPrecision?: number;
}

export interface IIdentifyRequest extends IAbstractRequest<'identify'> {
  geometry: Geometry;
  geometryPrecision?: number;
  geometryProjection: Projection;
  identifyTolerance?: number; // in pixel
  layersPrefix?: LayersPrefix;
  outFields?: string;
  returnFieldName?: boolean;
}

export interface IQueryResponse {
  request: IGisRequest;
  featureTypeResponses: IQueryFeatureTypeResponse[];
}

export interface IQueryFeatureTypeResponse {
  type: IFeatureType<any>;
  features: Feature[];
  source: IQuerySource;
}

export interface IAttribute {
  key: string;
  type?: 'Oid' | 'Boolean' | 'Number' | 'String' | 'Date' | 'Geometry' | 'Unknown';
  name?: string;
}

export interface IFeatureType<IDT extends number | string> {
  id: IDT;
  queryId?: IDT;
  hide?: boolean;
  name?: string;
  identifierAttribute?: IAttribute;
  attributes?: IAttribute[];
  readonly initialPredicate?: IPredicate;

  joinWithOr?: boolean;

  predicateWrapper?: FeatureTypePredicateWrapper;
}

export class FeatureTypePredicateWrapper {
  public joinWithOr: boolean;

  private _initialPredicate: IPredicate;
  private _additionalPredicates: Record<string, IPredicate> = {};

  public constructor(initialPredicate?: IPredicate, joinWithOr = true) {
    this._initialPredicate = initialPredicate;
    this.joinWithOr = joinWithOr;
  }

  public addPredicate(predicate: IPredicate): void {
    this._additionalPredicates[predicate.hashCode()] = predicate;
  }

  public removePredicate(hash: string) {
    delete this._additionalPredicates[hash];
  }

  get predicate(): IPredicate {
    return this.filterBuilder.predicate;
  }

  get filterBuilder(): FilterBuilder {
    const predicates: IPredicate[] = Object.values(this._additionalPredicates);
    if (predicates.length > 0) {
      let filterBuilder: FilterBuilder;
      for (const predicate of predicates) {
        if(!filterBuilder) {
          filterBuilder = new FilterBuilder(predicate);
        } else {
          if (this.joinWithOr) {
            filterBuilder.or(predicate);
          } else {
            filterBuilder.and(predicate);
          }
        }
      }
      return filterBuilder.and(this._initialPredicate);
    }
    return new FilterBuilder(this._initialPredicate);
  }

}
export type LayersPrefix = LayersPrefixEnum.ALL | LayersPrefixEnum.TOP | LayersPrefixEnum.VISIBLE;
export enum LayersPrefixEnum {
  ALL = 'all',
  TOP = 'top',
  VISIBLE = 'visible',
}
