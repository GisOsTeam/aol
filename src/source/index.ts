/**
 * @file Automatically generated by barrelsby.
 */

export * from './ExternalVector';
export * from './Image';
export * from './ImageArcGISRest';
export * from './ImageStatic';
export * from './ImageWms';
export * from './LocalVector';
export * from './QueryArcGISRest';
export * from './TileArcGISRest';
export * from './TileImage';
export * from './TileWms';
export * from './Vector';
export * from './VectorTile';
export * from './Wfs';
export * from './Xyz';
export * from './query/index';
export * from './types/index';
