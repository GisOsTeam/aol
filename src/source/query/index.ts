/**
 * @file Automatically generated by barrelsby.
 */

export * from './ags';
export * from './identify';
export * from './wms';
