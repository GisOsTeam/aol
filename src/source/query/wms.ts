import Feature from 'ol/Feature';
import { get as getProjection, transformExtent } from 'ol/proj';
import WMSGetFeatureInfo from 'ol/format/WMSGetFeatureInfo';
import {
  IGisRequest,
  IFeatureType,
  IQueryFeatureTypeResponse,
  IExtended,
  IAttribute,
  IIdentifyRequest,
} from '../IExtended';
import { toGeoJSONGeometry, disjoint } from '../../utils';
import { getForViewAndSize } from 'ol/extent';
import SimpleGeometry from 'ol/geom/SimpleGeometry';
import Projection from 'ol/proj/Projection';
import Geometry from 'ol/geom/Geometry';
import { HttpEngine } from '../../HttpEngine';

const format = new WMSGetFeatureInfo();

function getFeatureInfoOnBBOX(
  source: IExtended,
  type: IFeatureType<string>,
  queryType: 'query' | 'identify',
  requestProjectionCode: string,
  featureProjectionCode: string,
  bbox: number[],
  renderSize: number,
  tolerance: number,
  limit: number,
  id?: number | string,
  cql?: string
): Promise<Feature[]> {
  let url = '';
  if ('getUrl' in source) {
    url = (source as any).getUrl();
  } else if ('getUrls' in source) {
    url = (source as any).getUrls()[0];
  }
  const body: { [id: string]: string } = {};
  body.SERVICE = 'WMS';
  body.VERSION = '1.1.0';
  body.REQUEST = 'GetFeatureInfo';
  body.QUERY_LAYERS = type.id;
  body.LAYERS = type.id;
  body.INFO_FORMAT = 'application/vnd.ogc.gml';
  body.SRS = requestProjectionCode;
  body.X = `${Math.round(renderSize / 2)}`;
  body.Y = `${Math.round(renderSize / 2)}`;
  body.WIDTH = `${renderSize}`;
  body.HEIGHT = `${renderSize}`;
  body.BBOX = bbox.join(',');
  body.FEATURE_COUNT = `${limit}`;
  if (id != null) {
    body.FEATUREID = `${id}`; // GeoServer
    // ?? // QGis Server
    // ?? // MapServer
    // ?? // ArcGIS WMS
  }
  const toleranceStr = `${tolerance}`;
  body.BUFFER = toleranceStr; // GeoServer
  body.RADIUS = toleranceStr; // MapServer
  body.FI_POINT_TOLERANCE = toleranceStr; // QGis Server
  body.FI_LINE_TOLERANCE = toleranceStr; // QGis Server
  body.FI_POLYGON_TOLERANCE = toleranceStr; // QGis Server
  body.WITH_GEOMETRY = 'true'; // QGis Server
  if (cql != null && cql !== '') {
    body.CQL_FILTER = cql;
  }
  if (queryType === 'query') {
    body.SLD_BODY = `<StyledLayerDescriptor version="1.0.0"><UserLayer><Name>${type.id}</Name><UserStyle><FeatureTypeStyle><Rule><PointSymbolizer><Graphic><Mark><WellKnownName>square</WellKnownName><Fill><CssParameter name="fill">#FFFFFF</CssParameter></Fill></Mark><Size>1</Size></Graphic></PointSymbolizer><LineSymbolizer><Stroke><CssParameter name="stroke">#000000</CssParameter><CssParameter name="stroke-width">1</CssParameter></Stroke></LineSymbolizer><PolygonSymbolizer><Stroke><CssParameter name="stroke">#000000</CssParameter><CssParameter name="stroke-width">1</CssParameter></Stroke></PolygonSymbolizer></Rule></FeatureTypeStyle></UserStyle></UserLayer></StyledLayerDescriptor>`;
  }

  const httpEngine = HttpEngine.getInstance();
  return httpEngine
    .send({
      url,
      body,
      method: 'POST',
      contentType: 'application/x-www-form-urlencoded',
      responseType: 'text',
    })
    .then(
      (res) => {
        // Read features
        const features = [] as Feature[];
        // Search projection on results
        let dataProjection = getProjection(requestProjectionCode);
        let dataProjectionCode = requestProjectionCode;
        const res1 = res.body.match(/\ssrsName=\"([^\"]+)\"/i);
        if (res1 && res1.length >= 2) {
          const res2 = res1[1].match(/(\d+)(?!.*\d)/g);
          if (res2 && res2.length > 0) {
            dataProjectionCode = 'EPSG:' + res2[res2.length - 1];
          }
        }
        try {
          dataProjection = getProjection(dataProjectionCode);
        } catch (err) {
          console.error(err);
        }
        // Hack for GeoServer with space in name
        let txt = res.text;
        if (/\s/.test(type.id)) {
          const withoutSpace = type.id.replace(/\s/g, '_');
          const withSpace = type.id.replace(/([.*+?^=!:${}()|\[\]\/\\])/g, '\\$1');
          txt = txt.replace(new RegExp('<' + withSpace, 'g'), '<' + withoutSpace);
          txt = txt.replace(new RegExp('</' + withSpace, 'g'), '</' + withoutSpace);
        }
        // Read features
        const allFeatures = format.readFeatures(txt);
        if (allFeatures != null && allFeatures.length > 0) {
          allFeatures.forEach((feature: Feature) => {
            if (limit == null || features.length < limit) {
              if (dataProjection.getUnits() === 'degrees') {
                if (feature.getGeometry()) {
                  // In degree: This formats the geographic coordinates in longitude/latitude (x/y) order.
                  // Reverse coordinates !
                  (feature.getGeometry() as SimpleGeometry).applyTransform(
                    (input: number[], ouput: number[], dimension: number) => {
                      for (let i = 0; i < input.length; i += dimension) {
                        const y = input[i];
                        const x = input[i + 1];
                        ouput[i] = x;
                        ouput[i + 1] = y;
                      }
                      return ouput;
                    }
                  );
                }
              }
              if (feature.getGeometry()) {
                feature.getGeometry().transform(dataProjection, featureProjectionCode);
              }
              features.push(feature);
            }
          });
        }
        return features;
      },
      (err) => {
        console.error('Get WMS feature info in error');
        return err;
      }
    );
}

export function executeWmsQuery(
  source: IExtended,
  type: IFeatureType<string>,
  request: IGisRequest
): Promise<IQueryFeatureTypeResponse> {
  const { olMap, geometry, geometryProjection, queryType, limit } = request;
  const requestProjectionCode = 'EPSG:3857';
  const olView = olMap.getView();
  const mapProjection = olView.getProjection();
  const extent = transformExtent(geometry.getExtent(), geometryProjection, requestProjectionCode);
  const mapExtent = getForViewAndSize(
    [0.5 * extent[0] + 0.5 * extent[2], 0.5 * extent[1] + 0.5 * extent[3]],
    olView.getResolution(),
    0,
    [1001, 1001]
  );
  let tolerance;
  switch (queryType) {
    case 'identify':
      const { identifyTolerance } = request as IIdentifyRequest;
      if (Math.round(identifyTolerance) > 0) {
        tolerance = Math.round(identifyTolerance);
      } else {
        tolerance = 4;
      }
      break;
    case 'query':
      const mapH = Math.sqrt(
        (mapExtent[2] - mapExtent[0]) * (mapExtent[2] - mapExtent[0]) + (mapExtent[3] - mapExtent[1])
      );
      const geomH = Math.sqrt(
        (extent[2] - extent[0]) * (extent[2] - extent[0]) + (extent[3] - extent[1]) * (extent[3] - extent[1])
      );
      tolerance = 1 + Math.round((500 * geomH) / mapH);
      break;
  }

  return getFeatureInfoOnBBOX(
    source,
    type,
    queryType,
    requestProjectionCode,
    mapProjection.getCode(),
    mapExtent,
    1001,
    tolerance,
    limit ? 2 * limit : 100000
  ).then((allFeatures) => {
    const features = [] as Feature[];
    allFeatures.forEach((feature: Feature) => {
      // Check intersection
      if (
        feature.getGeometry() == null ||
        (queryType === 'identify' && (geometry.getType() === 'Point' || geometry.getType() === 'MultiPoint')) ||
        !disjoint(toGeoJSONGeometry(feature.getGeometry()), toGeoJSONGeometry(geometry))
      ) {
        features.push(feature);
      }
    });
    return {
      type,
      features,
      source,
    };
  });
}

export function retrieveWmsFeature(
  source: IExtended,
  type: IFeatureType<string>,
  id: number | string,
  featureProjection: Projection
): Promise<Feature> {
  const requestProjectionCode = 'EPSG:3857';
  const mapExtent = [-20026376.39, -20048966.1, 20026376.39, 20048966.1];
  return getFeatureInfoOnBBOX(
    source,
    type,
    'query',
    requestProjectionCode,
    featureProjection.getCode(),
    mapExtent,
    1001,
    500,
    1,
    id
  ).then((allFeatures) => {
    let feature = null;
    if (allFeatures != null && allFeatures.length > 0) {
      feature = allFeatures[0];
    }
    return feature;
  });
}

export function loadWmsFeatureDescription(source: IExtended, type: IFeatureType<string>): Promise<void> {
  const requestProjectionCode = 'EPSG:3857';
  const mapExtent = [-20026376.39, -20048966.1, 20026376.39, 20048966.1];
  return getFeatureInfoOnBBOX(
    source,
    type,
    'query',
    requestProjectionCode,
    requestProjectionCode,
    mapExtent,
    3,
    3,
    1
  ).then((allFeatures) => {
    let feature = null;
    if (allFeatures != null && allFeatures.length > 0) {
      type.attributes = [];
      feature = allFeatures[0];
      const properties = feature.getProperties();
      Object.keys(properties).forEach((key) => {
        const attribute: IAttribute = {
          key,
          type: 'Unknown',
        };
        const value = properties[key];
        if (value != null) {
          if (typeof value === 'string') {
            attribute.type = 'String';
          } else if (typeof value === 'object' && value instanceof Geometry) {
            attribute.type = 'Geometry';
          }
        }
        type.attributes.push(attribute);
      });
      return;
    }
  });
}
