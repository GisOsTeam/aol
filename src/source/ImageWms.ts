import OlImageWMS from 'ol/source/ImageWMS';
import {
  IQueryFeatureTypeResponse,
  IQueryRequest,
  IQueryResponse,
  IExtendedOptions,
  IFeatureType,
  IExtended,
} from './IExtended';
import { getWmsLayersFromTypes } from '../utils';
import { wmsQueryOne } from './query/wmsQuery';
import { LayerType, LayerTypeEnum } from './types/layerType';
import { SourceType, SourceTypeEnum } from './types/sourceType';
import { Options } from 'ol/source/ImageWMS';

export interface IImageWMSOptions extends IExtendedOptions, Options {
  types: IFeatureType<string>[];
}

export class ImageWms extends OlImageWMS implements IExtended {
  protected options: IImageWMSOptions;

  constructor(options: IImageWMSOptions) {
    super({ ...options } as any);
    this.options = { ...options };
    this.set('types', options.types);
    this.updateParams({ ...this.getParams(), LAYERS: getWmsLayersFromTypes(options.types) });
    this.on('propertychange', this.handlePropertychange);
  }

  public getSourceType(): SourceType {
    return SourceTypeEnum.ImageWms;
  }

  public getSourceOptions(): IImageWMSOptions {
    return this.options;
  }

  public setSourceOptions(options: IImageWMSOptions): void {
    this.options = { ...options };
    this.un('propertychange', this.handlePropertychange);
    this.set('types', options.types);
    this.updateParams({ ...this.getParams(), LAYERS: getWmsLayersFromTypes(options.types) });
    this.on('propertychange', this.handlePropertychange);
  }

  public getLayerType(): LayerType {
    return LayerTypeEnum.Image;
  }

  public isSnapshotable(): boolean {
    return this.options.snapshotable == null ? true : this.options.snapshotable; // true by default
  }

  public isListable(): boolean {
    return this.options.listable == null ? true : this.options.listable; // true by default
  }

  public query(request: IQueryRequest): Promise<IQueryResponse> {
    const promises: Promise<IQueryFeatureTypeResponse>[] = [];
    for (const type of this.options.types) {
      promises.push(wmsQueryOne(this, this.getUrl(), type, request));
    }
    return Promise.all(promises).then((featureTypeResponses: IQueryFeatureTypeResponse[]) => {
      return {
        request,
        featureTypeResponses,
      };
    });
  }

  private handlePropertychange = (event: any) => {
    const key = event.key;
    const value = event.target.get(key);
    if (key === 'types') {
      this.updateParams({ ...this.getParams(), LAYERS: getWmsLayersFromTypes(value) });
      this.options.types = value;
    }
  };
}
