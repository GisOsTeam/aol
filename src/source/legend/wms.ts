import { HttpEngine } from '../../HttpEngine';
import { srcToImage, getWmsLayersFromTypes } from '../../utils';
import { ILayerLegend, ILegendRecord, ILegendSource } from '../IExtended';

export function loadLegendWms(source: ILegendSource): Promise<ILegendRecord> {
  if (typeof (source as any).getLegendUrl === 'function') {
    const url = (source as any).getLegendUrl(undefined, { TRANSPARENT: true, SLD_VERSION: '1.1.0' });
    const key = getWmsLayersFromTypes((source as any).options.types);
    return srcToImage(url).then((image) => {
      return {
        [key]: [
          {
            image,
            srcImage: url,
            height: image.height,
            width: image.width,
          },
        ],
      };
    });
  }
  return Promise.reject(new Error('Source is not a WMS source'));
}
