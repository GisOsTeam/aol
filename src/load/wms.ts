import { createSource } from '../utils';
import { SourceTypeEnum } from '../source/types/sourceType';
import { IFeatureType } from '../source/IExtended';
import { ImageWms } from '../source/ImageWms';

/**
 * Load WMS.
 */
export function loadWMS(serverUrl: string, types: Array<IFeatureType<string>>, gisProxyUrl: string): Promise<ImageWms> {
  return new Promise<ImageWms>(resolve => {
    let url = serverUrl;
    if (gisProxyUrl != null && gisProxyUrl !== '') {
      url = `${gisProxyUrl}/${btoa(serverUrl)
        .replace('=', '%3D')
        .replace('/', '%2F')
        .replace('+', '%2B')}`;
    }
    const imageWms = createSource(SourceTypeEnum.ImageWms, { types, url }) as ImageWms;
    resolve(imageWms);
  });
}
