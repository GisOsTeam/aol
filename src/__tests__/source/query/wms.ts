import { executeWmsQuery, loadWmsFeatureDescription } from '../../../source/query/wms';
import OlMap from 'ol/Map';
import OlView from 'ol/View';
import { get as getProjection, Projection } from 'ol/proj';
import Polygon from 'ol/geom/Polygon';
import { IQueryFeatureTypeResponse, IGisRequest, IFeatureType } from '../../../source/IExtended';
import { ImageWms } from '../../../source/ImageWms';

const states = new ImageWms({
  url: 'https://download.data.grandlyon.com/wms/rdata',
  types: [{ id: 'metropole-de-lyon:adr_voie_lieu.adrbanc_latest' }],
  params: {},
});

test('describe wms', () => {
  const type: IFeatureType<string> = states.get('types')[0];
  return loadWmsFeatureDescription({
    url: states.getUrl() as string,
    type,
    method: 'GET',
    outputFormat: 'text/xml; subtype=gml/3.1.1',
    version: '1.3.0',
    requestProjectionCode: 'EPSG:3857',
  }).then(() => {
    if (!type.attributes) {
      throw new Error('Attributes should not be undefined');
    }
    expect<number>(type.attributes.length).toMatchSnapshot();
  });
});

test('query wms', () => {
  const bbox = [561829.47003365, 5747718.13304753, 561829.47003365, 5747718.13304753];
  const request: IGisRequest = {
    olMap: new OlMap({
      view: new OlView({
        center: [0, 0],
        zoom: 1,
      }),
    }),
    geometry: new Polygon([
      [
        [bbox[0], bbox[1]], // bas-gauche
        [bbox[0], bbox[3]], // haut-gauche
        [bbox[2], bbox[3]], // haut-droit
        [bbox[2], bbox[1]], // bas-droit
        [bbox[0], bbox[1]], // fermeture
      ],
    ]),
    geometryProjection: getProjection('EPSG:3857') as Projection,
    queryType: 'query',
  };
  const type: IFeatureType<string> = states.get('types')[0];
  return executeWmsQuery({
    source: states,
    url: states.getUrl() as string,
    type,
    request,
    method: 'GET',
    outputFormat: 'text/xml; subtype=gml/3.1.1', // 'application/json',
    version: '1.3.0',
    requestProjectionCode: 'EPSG:3857',
    swapXYBBOXRequest: false,
    swapLonLatGeometryResult: false,
  }).then((response: IQueryFeatureTypeResponse) => {
    expect<number>(response.features.length).toMatchSnapshot();
    expect(response.features[0].getProperties()).toMatchSnapshot();
  });
});

test('identify wms', () => {
  const bbox = [561829.47003365, 5747718.13304753, 561829.47003365, 5747718.13304753];
  const request: IGisRequest = {
    olMap: new OlMap({
      view: new OlView({
        center: [0, 0],
        zoom: 1,
      }),
    }),
    geometry: new Polygon([
      [
        [bbox[0], bbox[1]], // bas-gauche
        [bbox[0], bbox[3]], // haut-gauche
        [bbox[2], bbox[3]], // haut-droit
        [bbox[2], bbox[1]], // bas-droit
        [bbox[0], bbox[1]], // fermeture
      ],
    ]),
    geometryProjection: getProjection('EPSG:3857') as Projection,
    queryType: 'identify',
  };
  const type: IFeatureType<string> = states.get('types')[0];
  return executeWmsQuery({
    source: states,
    url: states.getUrl() as string,
    type,
    request,
    method: 'GET',
    outputFormat: 'text/xml; subtype=gml/3.1.1', // 'application/json',
    version: '1.3.0',
    requestProjectionCode: 'EPSG:3857',
    swapXYBBOXRequest: false,
    swapLonLatGeometryResult: false,
  }).then((response: IQueryFeatureTypeResponse) => {
    expect<number>(response.features.length).toMatchSnapshot();
    expect(response.features[0].getProperties()).toMatchSnapshot();
  });
});
