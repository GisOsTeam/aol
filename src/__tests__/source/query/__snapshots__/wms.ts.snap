// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`describe wms 1`] = `6`;

exports[`identify wms 1`] = `10509`;

exports[`identify wms 2`] = `
{
  "boundedBy": [
    534333.55580771,
    5732403.6615682,
    534333.55580771,
    5732403.6615682,
  ],
  "dossier": "true",
  "gid": "1",
  "insee": "69387",
  "source": "Open Street Map",
  "the_geom": Point {
    "dispatching_": {},
    "disposed": false,
    "eventTarget_": undefined,
    "extentRevision_": -1,
    "extent_": [
      Infinity,
      Infinity,
      -Infinity,
      -Infinity,
    ],
    "flatCoordinates": [
      534333.55580771,
      5732403.6615682,
      0,
    ],
    "layout": "XYZ",
    "listeners_": {
      "change": [
        [Function],
      ],
    },
    "ol_uid": "79",
    "on": [Function],
    "once": [Function],
    "pendingRemovals_": {},
    "revision_": 2,
    "simplifiedGeometryMaxMinSquaredTolerance": 0,
    "simplifiedGeometryRevision": 0,
    "simplifyTransformedInternal": [Function],
    "stride": 3,
    "un": [Function],
    "values_": null,
  },
}
`;

exports[`query wms 1`] = `6`;

exports[`query wms 2`] = `
{
  "boundedBy": [
    561829.47003365,
    5747718.13304753,
    561829.47003365,
    5747718.13304753,
  ],
  "dossier": "true",
  "gid": "1191",
  "insee": "69279",
  "source": "Open Street Map",
  "the_geom": Point {
    "dispatching_": {},
    "disposed": false,
    "eventTarget_": undefined,
    "extentRevision_": -1,
    "extent_": [
      Infinity,
      Infinity,
      -Infinity,
      -Infinity,
    ],
    "flatCoordinates": [
      561829.47003365,
      5747718.13304753,
      0,
    ],
    "layout": "XYZ",
    "listeners_": {
      "change": [
        [Function],
      ],
    },
    "ol_uid": "33",
    "on": [Function],
    "once": [Function],
    "pendingRemovals_": {},
    "revision_": 2,
    "simplifiedGeometryMaxMinSquaredTolerance": 0,
    "simplifiedGeometryRevision": 0,
    "simplifyTransformedInternal": [Function],
    "stride": 3,
    "un": [Function],
    "values_": null,
  },
}
`;
